File: src/comms.c
/*===================================================/
/  Comms routine for PX425 Assignment 5. Some basic  /
/  routines/ideas lifted from Assignment 4 code.     /
/  All interactions with MPI libraries are in here.  /
/                                                    /
/  Original code by C. Woodgate                      /
/  (based on code by N. Hine, D. Quigley)            /
/===================================================*/
#include "mpi.h"
#include <stdio.h>
#include <stdlib.h>
#include <math.h>
#include "trajectories.h"
#include "comms.h"

int p;               /* Number of processor       */
int my_rank;         /* Rank of current processor */

MPI_Comm comm = MPI_COMM_WORLD;  /* Cartesian communicator    */

double t1,t2;		/* Time of initialisation and shutdown */

/*============/
/  FUNCTIONS  /
/============*/

void comms_wait()
{
	/*==================================================================/
	/ Function to allow me to call MPI_Barrier to sync threads.         /
	/==================================================================*/
	
	/* Call MPI_Barrier */
	MPI_Barrier(comm);
	return;
}

void comms_initialise(int argc, char **argv) 
{
	/*==================================================================/
	/ Function to initialise MPI, get the communicator size p and the   /
	/ rank my_rank of the current process within that communicator.     /
	/==================================================================*/

	// Initialise MPI 
	MPI_Init(&argc, &argv);

	// Get the rank of the current process
	MPI_Comm_rank(comm, &my_rank);

	// Get the size of the current communicator
	MPI_Comm_size (comm, &p);

	// Start the timer
	t1 = MPI_Wtime();

	return;
}

void comms_phase_space_local(int my_rank, int nFragments, int n_local)
{
	/*==================================================================/
	/ Function to set up the phase space for the system on each thread  /
	/ in 'local_' arrays from rank 0, where we called explodeTheMoon(). /
	/==================================================================*/
	
	//Number of elements to be sent when vectors
	int n_send = 3*n_local;
	
	// Broadcast positions array to all ranks
	MPI_Bcast(pos, nFragments*3, MPI_DOUBLE, 0, comm);

	//Boradcast mass information to all ranks
	MPI_Bcast(mass, nFragments, MPI_DOUBLE, 0, comm);
	
	//Give each rank information about whether fragment is a dummy or not
	MPI_Bcast(calc_active, nFragments, MPI_INT, 0, comm);
	
	// Send each rank the section of the velocity array on which they will work
	MPI_Scatter(global_vel, n_send, MPI_DOUBLE, local_vel, n_send, MPI_DOUBLE, 0, comm);

	// Send each rank their section of the acceleration array.
	// (Almost redundant but as we only call it once doesn't cost too much)
	MPI_Scatter(global_acc, n_send, MPI_DOUBLE, local_acc, n_send, MPI_DOUBLE, 0, comm);

	// Send each rank their relevant radii
	MPI_Scatter(global_radius, n_local, MPI_DOUBLE, local_radius, n_local, MPI_DOUBLE, 0, comm);

	// Send each rank the 'active' array
	MPI_Scatter(global_active, n_local, MPI_INT, local_active, n_local, MPI_INT, 0, comm);
	
	return;
}

void comms_positions_gather(int my_rank, int n_alloc, int n_local)
{
	/*==================================================================/
	/ Function to gather the positions array onto each thread after     /
	/ each timestep using MPI_Allgather.                                /
	/==================================================================*/

	// Number of elements to be sent
	int n_send = 3*n_local;

	// Wait until all threads have finished their calculations
	MPI_Barrier(comm);

	// Gather positions array on all threads
	MPI_Allgather(local_pos, n_send, MPI_DOUBLE, pos, n_send, MPI_DOUBLE, comm); 

	return;
}

void comms_impacts(int n_local)
{
	/*==================================================================/
	/ Function to communicate information needed about impacts on       /
	/ each thread.                                                      /
	/==================================================================*/

	// Work out how many fragments are active in each section	
	MPI_Reduce(&local_nActive, &nActive, 1, MPI_INT, MPI_SUM, 0, comm);

	// Work out which fragment is closest overall
	MPI_Reduce(&local_minRadsq, &minRadsq, 1, MPI_DOUBLE, MPI_MIN,0, comm);

	// Gather the local_active arrays onto thread zero
	MPI_Gather(local_active, n_local, MPI_INT, global_active, n_local, MPI_INT, 0, comm);
	return;
}

void comms_finalise() 
{
	/*===========================================================/
	/ Function to finalise MPI functionality and exit cleanly    /
	/===========================================================*/
  
	// Measure the time t2 using MPI_Wtime()
	t2 = MPI_Wtime();

	if (my_rank==0) 
	{
    	printf("Total time elapsed since MPI initialised :  %12.6f s\n",t2-t1);
	}

	// Shutdown MPI
	MPI_Finalize();

	return;
}

File: src/comms.f90
!----------------------------------------------------------------------!
! comms.f90                                                            !
!                                                                      !
! Module for handling calls to MPI library (used for parallelisation). !
!                                                                      !
! C. D. Woodgate,  Warwick                                        2023 !
!----------------------------------------------------------------------!
module comms

  use mpi
  use kinds
  use mpi_shared_data
  use io

  implicit none

  save

  ! total number of processes
  integer :: p

  ! rank of my processor
  integer :: my_rank

  ! start and end times for mpi
  real(real64) :: t1, t2

  ! mpi status_info
  integer, dimension(mpi_status_size) :: status_info

  ! error variables
  integer :: ierr

  ! our communicator
  integer :: cart_comm

  ! this processor coordinates
  integer, dimension(3) :: my_rank_coords

  ! neighbouring ranks
  integer, dimension(6) :: my_rank_neighbours
  integer :: east, west, north, south, up, down

  contains


  !--------------------------------------------------------------------!
  ! Routine to initialise MPI                                          !
  !                                                                    !
  ! C. D. Woodgate,  Warwick                                      2023 !
  !--------------------------------------------------------------------!
  subroutine comms_initialise()

    ! initialise mpi
    call mpi_init(ierr)

    ! set up the size and rank of the communicator
    call mpi_comm_rank(mpi_comm_world,my_rank,ierr)
    call mpi_comm_size(mpi_comm_world,p,ierr)

    ! start the clock
    t1 = mpi_wtime()

  end subroutine comms_initialise

  !--------------------------------------------------------------------!
  ! Routine to put in a call to MPI_BARRIER (wait for all processes).  !
  !                                                                    !
  ! C. D. Woodgate,  Warwick                                      2023 !
  !--------------------------------------------------------------------!
  subroutine comms_wait()

    ! Call MPI_BARRIER
    call mpi_barrier(mpi_comm_world, ierr)

  end subroutine comms_wait

  !--------------------------------------------------------------------!
  ! Routine to reduce results of an (MPI) ensemble of simulations and  !
  ! perform averaging.                                                 !
  !                                                                    !
  ! C. D. Woodgate,  Warwick                                      2023 !
  !--------------------------------------------------------------------!
  subroutine comms_reduce_results(setup)

    ! Input contains information about simulation
    type(run_params), intent(in) :: setup

    ! Bring all simulation energy arrays, <E>(T), to rank 0
    ! and sum them.
    call mpi_reduce(energies_of_T, av_energies_of_T, setup%T_steps,    &
                    MPI_DOUBLE, MPI_SUM, 0, mpi_comm_world, ierr)

    ! Divide by the number of simulations to get the average
    av_energies_of_T = av_energies_of_T/real(p)
  
    ! Do the same for the heat capacity data
    call mpi_reduce(C_of_T, av_C_of_T, setup%T_steps,                  &
                 MPI_DOUBLE_PRECISION, MPI_SUM, 0, mpi_comm_world, ierr)
    av_C_of_T = av_C_of_T/real(p)
  
    ! Do the same with the acceptance rates
    call mpi_reduce(acceptance_of_T, av_acceptance_of_T, setup%T_steps,&
                 MPI_DOUBLE_PRECISION, MPI_SUM, 0, mpi_comm_world, ierr)
    av_acceptance_of_T = av_acceptance_of_T/real(p)
  
    ! Do the same with the radial densities
    call mpi_reduce(rho_of_T, av_rho_of_T,                             &
                    setup%T_steps*(setup%n_species**2)*setup%wc_range, &
                    MPI_DOUBLE_PRECISION, MPI_SUM, 0, mpi_comm_world,  &
                    ierr)
    av_rho_of_T = av_rho_of_T/real(p)

  end subroutine comms_reduce_results

  !--------------------------------------------------------------------!
  ! Routine to finalise MPI and display time taken.                    !
  !                                                                    !
  ! C. D. Woodgate,  Warwick                                      2023 !
  !--------------------------------------------------------------------!
  subroutine comms_finalise()

    ! stop the clock
    t2 = mpi_wtime()

!    ! print time taken
!    if(my_rank .eq. 0) then
!      print '(a,f0.2,a)', "Total time elapsed for mpi run is ", t2-t1, &
!                          " seconds."
!    end if

    ! clean up mpi
    call mpi_finalize(ierr)

  end subroutine comms_finalise

end module comms
